### DISCUSSION TITLE
Launcher multiple tcp server?

### DISCUSSION THREAD

**Discussion by azzahrah** [activej/activej#172] [Q&A]
Launcher
```
public class TrackerServerLauncher extends Launcher {

    public static final int PORT = 8080;
    public static final int WORKERS = 4;

    public String PROPERTIES_FILE = "";

    public TrackerServerLauncher(String file) {
        this.PROPERTIES_FILE = file;
    }

    @Inject
    PrimaryServer primaryServer;

    @Provides
    Eventloop primaryEventloop(Config config) {
        return Eventloop.create()
                .withInitializer(ofEventloop(config.getChild("eventloop.primary")));
    }

    @Provides
    @Worker
    Eventloop workerEventloop(Config config, @Optional ThrottlingController throttlingController) {
        return Eventloop.create()
                .withInitializer(ofEventloop(config.getChild("eventloop.worker")))
                .withInitializer(eventloop -> eventloop.withInspector(throttlingController));
    }

    @Provides
    @Worker
    Consumer<AsyncTcpSocket> socketConsumer(@WorkerId int workerId) {
        return s -> {
            System.out.println("Device GT06 Connected:");
            //processing incoming data
        };
    }

    @Provides
    @Worker
    SimpleServer workerServer(Eventloop eventloop, Consumer<AsyncTcpSocket> socketConsumer, Config config) {
        SimpleServer ss = SimpleServer.create(eventloop, socketConsumer);
        try {
            ss.listen();
        } catch (IOException ex) {
            ex.printStackTrace();
        }
        return ss;
    }

    @Provides
    WorkerPool workerPool(WorkerPools workerPools, Config config) {
        return workerPools.createPool(config.get(ofInteger(), "workers", WORKERS));
    }

    @Provides
    PrimaryServer primaryServer(Eventloop primaryEventloop, WorkerPool.Instances<SimpleServer> workerServers, Config config) {
        return PrimaryServer.create(primaryEventloop, workerServers.getList())
                .withInitializer(ofPrimaryServer(config.getChild("tcp")));
    }

    @Provides
    Config config() {
        Config c = Config.create()
                .with("tcp.listenAddresses", Config.ofValue(ofInetSocketAddress(), new InetSocketAddress(PORT)))
                .with("workers", "" + WORKERS)
                .overrideWith(ofClassPathProperties(PROPERTIES_FILE, true))
                .overrideWith(ofSystemProperties("config"));
        System.out.println(c.toProperties().toString());
        return c;
    }

    @Override
    protected final io.activej.inject.module.Module getModule() {
        return combine(
                ServiceGraphModule.create(),
                WorkerPoolModule.create(),
                JmxModule.create()
                        .withInitializer(ofGlobalEventloopStats()),
                ConfigModule.create()
                        .withEffectiveConfigLogger()
        );
    }

    @Override
    protected void run() throws Exception {
        awaitShutdown();
    }

}
```

Main App

```
public class MultithreadTcpServer {

    public static void main(String[] args) throws Exception {
      //start server1 port 8081
       new Thread(()->{
        String file = "gt06.properties";
        TrackerServerLauncher server1 = new TrackerServerLauncher(file);
        try {
            server1.launch(new String[]{});
        } catch (Exception ex) {
            Logger.getLogger(MultithreadTcpServer.class.getName()).log(Level.SEVERE, null, ex);
        }
      }).start();

   //start server2 port 8082
       new Thread(()->{
        String file2 = "coban.properties";
        TrackerServerLauncher server2 = new TrackerServerLauncher(file2);
        try {
            server2 .launch(new String[]{});
        } catch (Exception ex) {
            Logger.getLogger(MultithreadTcpServer.class.getName()).log(Level.SEVERE, null, ex);
        }
      }).start();
    }
}
```

why only second server started (port 8082)? mybe any sample howto launch multiple tcp server with primaryserver+worker style?

**Comment by eduard-vasinskyi** [comment]
Hello, @azzahrah
A [Launcher](https://github.com/activej/activej/blob/master/launcher/src/main/java/io/activej/launcher/Launcher.java) is essentially a complete application. You only need to launch one Launcher per application. If you need to have two listening servers, both with their own set of worker servers, you can use multiple [WorkerPool](https://github.com/activej/activej/blob/master/boot-workers/src/main/java/io/activej/worker/WorkerPool.java)s.

Here is a [gist](https://gist.github.com/eduard-vasinskyi/a9ec608e1ea7effb6f11281e7704e658) that demonstrates just that.

Notice that we provide 2 named [WorkerPool](https://github.com/activej/activej/blob/master/boot-workers/src/main/java/io/activej/worker/WorkerPool.java)s (`“Coban”` and `“GT06”`). These pools can have different amounts of workers, depending on your needs. We also provide 2 named primary [Eventloop](https://github.com/activej/activej/blob/master/core-eventloop/src/main/java/io/activej/eventloop/Eventloop.java)s for both servers as well as primary servers themselves. We do not change [@Worker](https://github.com/activej/activej/blob/master/boot-workers/src/main/java/io/activej/worker/annotation/Worker.java) bindings - they will be used by both WorkerPools.

In a gist we used one `‘.properties’` file for configuration. Configurations for both servers are prefixed with either _‘coban’_ or _‘gt06’_ like this: 
 - `coban.http.listenAdrresses=*:9001`
 - `gt06.http.listenAddresses=*:9002`

Also, you may take a look at generated graphvizs for DI and service graphs (printed to stdout when you launch your application) to analyze your setup:
![image](https://user-images.githubusercontent.com/39034539/122538188-f5739d80-d02e-11eb-90c4-75e18f8f8912.png)
Here you can see that there are 2 primary servers, each has its own primary `Eventloop`. Each primary server has 4 worker servers, each with a dedicated worker `Eventloop`

By the way, you do not need to manually call `listen()` on a provided server. Since you use [ServiceGraphModule](https://github.com/activej/activej/blob/master/boot-servicegraph/src/main/java/io/activej/service/ServiceGraphModule.java) in your `Launcher`, it will handle start/stop routines for each service.
Also, you may look at the [ComplexHttpLauncher](https://github.com/activej/activej/blob/master/launchers/http/src/test/java/io/activej/launchers/http/ComplexHttpLauncher.java) example, which has multiple worker pools, custom worker scopes, etc.

**Comment by azzahrah** [comment]
@eduard-vasinskyi  thank that really cool....

**Comment by azzahrah** [comment]
Helo @eduard-vasinskyi 

i just modified my code like this:
i have two or more instance class abastractModule


Module Concox
```
public class ModuleConcox extends AbstractModule {


    
    ConfigServer configServer;
    public ModuleConcox(ConfigServer configServer){
        this.configServer=configServer;
    }

    @Provides
    @WorkerGT06
    Eventloop myWorkerEventloop() {
        return Eventloop.create();
    }

    @Provides
    @WorkerGT06
    PrimaryServer primaryServer(Eventloop primaryEventloop,WorkerPool.Instances<ServerTracker> workerServers) {
        return PrimaryServer.create(primaryEventloop, workerServers.getList())
                .withListenPort(9003);
    }


    @Provides
    @WorkerGT06
    WorkerPool workerPoolGT06(WorkerPools workerPools) {
        return workerPools.createPool(WORKERS);
    }
    @Provides
    @WorkerGT06
    ServerTracker workerServer(Eventloop eventloop, @WorkerId int workerId) {
        return new ServerTracker(workerId,eventloop,configServer,new DecoderConcoxFactory());
    }
}

```
Module TLT
```
public class ModuleTLT extends AbstractModule {


    ConfigServer configServer;
    public ModuleTLT(ConfigServer configServer){
        this.configServer=configServer;
    }

     @Provides
    @WorkerTLT
    Eventloop myWorkerEventloop() {
        return Eventloop.create();
    }
    @Provides
    @WorkerTLT
    PrimaryServer primaryServer(Eventloop primaryEventloop,WorkerPool.Instances<ServerTracker> workerServers) {
        return PrimaryServer.create(primaryEventloop, workerServers.getList())
                .withListenPort(9004);
    }
    @Provides
    @WorkerTLT
    WorkerPool workerPoolCoban(WorkerPools workerPools) {
        return workerPools.createPool(WORKERS);
    }
    @Provides
    @WorkerTLT
    ServerTracker workerServer(Eventloop eventloop, @WorkerId int workerId) {
        return new ServerTracker(workerId,eventloop,configServer,new DecoderTLTFactory());
    }
}

```
Inside Launcher
```
   public class App extends Launcher {
    @Inject
    @WorkerGT06
    PrimaryServer primaryServerConcox;

    @Inject
    @WorkerTLT
    PrimaryServer primaryServerTLT;


   @Override
    protected final Module getModule() {
        return combine(
                ServiceGraphModule.create(),
                WorkerPoolModule.create(WorkerGT06.class, WorkerTLT.class),
                new ModuleConcox(new ConfigServer(9003)),
                new ModuleTLT(new ConfigServer(9004))
        );
    }
```

but we go something error like this:
```
Exception in thread "main" io.activej.inject.binding.DIException: Unsatisfied dependencies detected:
	key PrimaryServer required to make:
		- InstanceInjector<App> at <unknown binding location>
```


another question, howto setup like this using binding and Inject manually , to learn more about binding and inject in my case..

**Comment by eduard-vasinskyi** [comment]
Hi, @azzahrah

There are many ways to do it, depending on what you need.
1. If you want to create two different worker pools with separate eventloops and even number of workers, you can create two scope annotations as you did in your snippet.
2. If you want to reuse the same worker pool for both of your servers and eventloops, you can just use `@Named` annotation.

We have added examples for both cases ([a single pool example](https://github.com/activej/activej/blob/master/examples/core/http/src/main/java/doubleservers/DoubleServersSinglePool.java) and [a double pool example](https://github.com/activej/activej/blob/master/examples/core/http/src/main/java/doubleservers/DoubleServersTwoPools.java)). 

Those examples are somewhat similar to the snippet that you sent, although we use [AsyncHttpServer](https://github.com/activej/activej/blob/master/core-http/src/main/java/io/activej/http/AsyncHttpServer.java) there rather than a custom server.

We have also added [an example](https://github.com/activej/activej/blob/master/examples/core/http/src/main/java/doubleservers/DoubleServersTwoPoolsManual.java) of using manual binds/injects for a double pool example.
If you have any questions about those examples or ActiveJ in general, please let me know.

**Comment by azzahrah** [comment]
@eduard-vasinskyi, thank very much, its really complete sample, i will try and let you know if have same question..
